File: Lab2/program/function.cpp
#include "function.hpp"
#include <iostream>
#include <fstream>
#include <string>
#include <stdio.h>
#include <vector>

using namespace std;



string to_do_list(){
    string name;
    EList text;
    string task_name, time_start, duration = "0", preTime="";

    
//  вводимо імʼя файлу
    cout << "Enter input file: ";
    cin >> name;
    name = name + ".dat";
    
    ofstream fout(name, ios::binary);

    char new_line = 'y';
    while(new_line == 'y'){
        cout<<"Enter a name of task(maximum number of characters - 20): ";
        cin>>task_name;
        task_name = verifyName(task_name);
        text.name = task_name;
        
        cout<<"Enter an event start time in this format HH:MM: ";
        cin>>time_start;
        time_start = verifyTime(time_start, duration, preTime);
        preTime = time_start;
        text.time_start = time_start;
        
        
        cout<<"Enter a duration in this format HH:MM: ";
        cin>>duration;
        duration = verifyDuration(duration, time_start);
        text.duration = duration;
        
        
        fout.write((char*)&text, sizeof(EList));

        if(duration_normal(duration, time_start)){
            cout << "\nDo you want to continue input task?[y/n]: ";
            cin >> new_line;
        }
        else{
            new_line = 'n';
        }
        cout<<endl;
    }


    fout.close();

    return name;
}


string verifyName(string name){
    while (name.size()>20 || name.size()<1){
        cout << "Enter a name again: ";
        cin >> name;
    }
    return name;
}

string verifyTime(string time, string duration, string preTime){
    bool flag_hours = hours(time);
    bool flag_duration = dur(time, duration, preTime);

    while(time.size()!=5 || time[2]!=':' || flag_hours || flag_duration){
        cout << "Enter a time again: ";
        cin >> time;
        flag_hours = hours(time);
        flag_duration = dur(time, duration, preTime);
    }
    return time;
}

string verifyDuration(string duration, string time){
    bool flag_hours = hours(duration);
    bool flag_dif = false;
    if(time_to_int(duration)+time_to_int(time) > 1440) flag_dif = true;
    while(duration.size()!=5 || duration[2]!=':' || flag_hours || flag_dif){
        cout << "Enter a duration again: ";
        cin >> duration;
        flag_hours = hours(duration);
        flag_dif = false;
        if(time_to_int(duration)+time_to_int(time) > 1440) flag_dif = true;
    }
    return duration;
}

bool hours(string time){
    bool flag = true;
    string h, m;
    h = time.substr(0, 2);
    m = time.substr(3, 2);
    int number_h = atoi(h.c_str()), number_m = atoi(m.c_str());

    if(number_h>=0 || number_h<=23){
        if(number_m>=0 || number_m<=59){
            flag = false;
        }
    }
    return flag;
}

bool dur(string time, string duration, string preTime){
    bool flag = false;
    int number = time_to_int(time);
    if(preTime!=""){
        int time_before = time_to_int(preTime);
        int d = time_to_int(duration);
        if(time_before + d > number){
            flag = true;
        }
    }
    return flag;
}

int time_to_int(string time){
    string h, m;
    h = time.substr(0, 2);
    m = time.substr(3, 2);
    int number = atoi(h.c_str())*60 + atoi(m.c_str());
    return number;
}

bool duration_normal(string duration, string time){
    bool flag = true;
    if(time_to_int(duration) + time_to_int(time) == 1440) flag = false;
    return flag;
}

void output_first(string name){//виводимо текст файлу
    EList text;
    ifstream fin(name, ios::binary);
    
    cout<<"File "<< name <<": \n";
    
    while (fin.read((char*)&text, sizeof(EList))){
        cout<< text.name <<", "<<  int_to_time(text.time_start)<<", "<<"duration: "<< int_to_time(text.duration)<< ";"<<endl;
    }
    fin.close();
    cout<<endl;
}

void output_second(string name){//виводимо текст файлу
    FreeList text;
    ifstream fin(name, ios::binary);
    
    cout<<"File "<< name <<": \n";
    
    while (fin.read((char*)&text, sizeof(FreeList))){
        cout<< int_to_time(text.start) <<"-"<< int_to_time(text.finish)<<", "<<"duration: "<< int_to_time(text.free_dur)<< ";"<<endl;
    }
    
    fin.close();
    cout<<endl;
}

vector <EList> file_to_text(string name){
    vector<EList> pack;
    EList text;
    ifstream fin(name, ios::binary);
    
    while (fin.read((char*)&text, sizeof(EList))){
        pack.push_back(text);
    }
    fin.close();
    cout<<endl;
    
    return pack;
}

void next_event(vector <EList> text){
    int j;
    int hour_task, minute_task;
    time_t rawtime;
    int time_now, time_task;
    int dif, min = 1440;
    string m;
    struct tm * timeinfo;
    time( &rawtime );
    timeinfo = localtime ( &rawtime );
    int hour = timeinfo->tm_hour, minute = timeinfo->tm_min;
    if(hour>=10 && minute<10) cout<<"Time now: "<< hour << ":" << "0"<< minute <<endl;
    if(hour<10 && minute<10) cout<<"Time now: "<< "0" << hour << ":" << "0"<< minute <<endl;
    if(hour<10 && minute>=10) cout<<"Time now: "<< "0" << hour << ":" << minute <<endl;
    if(hour>=10 && minute>=10) cout<<"Time now: "<< hour << ":" << minute <<endl;
    time_now = hour * 60 + minute;
    for(int i = 0; i < text.size(); i++){
        hour_task = text[i].time_start.hours;
        minute_task = text[i].time_start.minutes;
        time_task = hour_task * 60 + minute_task;
        dif = time_task - time_now;
        if(dif>0 && dif<min){
            j = i;
            min = dif;
        }
    }
    if(dif>0){
        cout<<"Next event: "<< text[j].name <<", "<<  int_to_time(text[j].time_start)<<", "<<"duration: "<< int_to_time(text[j].duration)<< ";"<<endl<<endl;
    }
    else cout<<"Today you have no more business!!!"<<endl<<endl;
}

string free_time(vector <EList> text){
    FreeList file;
    string s, i;
    int time_n, dur_n;
    string start = "13:00", finish = "23:59", free_dur = "10:59";
    bool task_flag = false;
    string name2;
    cout<<"Enter output file: "; cin>>name2;//називаємо другий файл

    name2+=".dat";
    ofstream fout(name2, ios::binary);
    
    for(int i = 0; i < text.size(); i++){
        time_n = text[i].time_start.hours * 60 + text[i].time_start.minutes;
        dur_n = text[i].duration.hours * 60 + text[i].duration.minutes;
        if(time_n + dur_n>=780){
            start = int_to_time(time_n + dur_n);
            if(i + 1 < text.size()){
                finish = int_to_time(text[i+1].time_start.hours * 60 + text[i+1].time_start.minutes);
            }
            else finish = "23:59";
            free_dur = int_to_time(time_to_int(finish) - time_to_int(start));
            file.start = start;
            file.finish = finish;
            file.free_dur = free_dur;
            fout.write((char*)&file, sizeof(FreeList));
            task_flag = true;
        }
    }
    if(!task_flag){
        file.start = start;
        file.finish = finish;
        file.free_dur = free_dur;
        fout.write((char*)&file, sizeof(FreeList));
    }
    
    fout.close();
    
    return name2;
        
}

string int_to_time(int num){
    string time;
    int h, m;
    h = num/60;
    m = num%60;
    if(h<10 && m<10) time ="0" + to_string(h) + ":" + "0" + to_string(m);
    if(m<10 && h>=10) time = to_string(h) + ":" + "0" + to_string(m);
    if(h<10 && m>=10) time = "0" + to_string(h) + ":" + to_string(m);
    if(h>=10 && m>=10) time = to_string(h) + ":" + to_string(m);
    return time;
}

string int_to_time(Time num){
    string time;
    int h, m;
    h = num.hours;
    m = num.minutes;
    if(h<10 && m<10) time ="0" + to_string(h) + ":" + "0" + to_string(m);
    if(m<10 && h>=10) time = to_string(h) + ":" + "0" + to_string(m);
    if(h<10 && m>=10) time = "0" + to_string(h) + ":" + to_string(m);
    if(h>=10 && m>=10) time = to_string(h) + ":" + to_string(m);
    return time;
}
       

File: Lab2/program/function.py
from datetime import datetime
#import numpy as np
import pickle

class Time:

    minutes = 0
    hours = 0

    def createTime(args):
        time = Time()
        time.hours = int(args[:2])
        time.minutes = int(args[3:])

        return time

    def toString(self):

        if self.hours < 10 and self.minutes < 10:
            time = "0" + str(self.hours) + ":" + "0" + str(self.minutes)
        elif self.minutes < 10 and self.hours >= 10:
            time = str(self.hours) + ":" + "0" + str(self.minutes)
        elif self.hours < 10 and self.minutes >= 10:
            time = "0" + str(self.hours) + ":" + str(self.minutes)
        elif self.hours >= 10 and self.minutes >= 10:
            time = str(self.hours) + ":" + str(self.minutes)
        return time

    def __init__(self):
        self.minutes = int()
        self.hours = int()


class Task:
    name = str()
    time_start = Time()
    duration = Time()

    def __init__(self,name,time_start,duration):
        self.name = name
        self.time_start = Time.createTime(time_start)
        self.duration = Time.createTime(duration)


class FreeT:
    start = Time()
    finish = Time()
    free_dur = Time()

    def __init__(self,start,finish,free_dur):
        self.start = Time.createTime(start)
        self.finish = Time.createTime(finish)
        self.free_dur = Time.createTime(free_dur)



def time_to_int(time: str):
    h = time[:2]
    m = time[3:]
    number = int(h) * 60 + int(m)
    return number

def hours(time: str):
    flag = True
    h = time[:2]
    m = time[3:]
    if int(h) >= 0 and int(h) <= 23:
        if int(m) >= 0 and int(m) <= 59:
            flag = False

    return flag

def dur(time: str, duration: str, times: str):
    flag = False
    number = time_to_int(time)
    if len(time) != 0:
        if len(times) != 0:
            num_t = len(times)-1
            time_before = time_to_int(times[num_t])
            d = time_to_int(duration)
            if time_before + d > number:
                flag = True
    return flag

def verifyName(name: str):
    while (len(name) > 20 or len(name) < 1):
        print("Enter a name again: ")
        name = input();
    return name

def verifyTime(time: str, duration, times):
    flag_hours = hours(time)
    flag_duration = dur(time, duration, times)
    while not (len(time) == 5 and time[2] == ':' and not flag_hours and not flag_duration):
        print("Enter a time again: ")
        time = input()
        flag_hours = hours(time)
        flag_duration = dur(time, duration, times)

    return time

def verifyDuration(duration: str, time: str):
    flag_hours = hours(duration)
    flag_dif = False
    if time_to_int(duration) + time_to_int(time) > 1440:
        flag_dif = True
    while len(duration) != 5 or duration[2] != ':' or flag_hours or flag_dif:
        print('Enter a duration again: ')
        duration = input()
        flag_hours = hours(duration)
        flag_dif = False
        if time_to_int(duration) + time_to_int(time) > 1440:
            flag_dif = True
    return duration

def duration_normal(duration: str, time: str):
    flag = True
    if time_to_int(duration) + time_to_int(time) == 1440:
        flag = False
    return flag

def int_to_time(num: int):
    h = num // 60
    m = num % 60
    if h < 10 and m < 10:
        time ="0" + str(h) + ":" + "0" + str(m);
    elif m < 10 and h >= 10:
        time = str(h) + ":" + "0" + str(m);
    elif h < 10 and m >= 10:
        time = "0" + str(h) + ":" + str(m);
    elif h >= 10 and m >= 10:
        time = str(h) + ":" + str(m);
    return time;


def capture_text(name_f):
    """
    Читаємо текст строками
    """
    flag = 'y'
    duration = '0'
    times = []
    # durations = []
    # text = np.array([])
    # text = []
    size = 0
    #tasks = Task()
    with open(name_f, 'wb') as file:
        while flag == 'y':
            print('Enter a name of task(maximum number of characters - 20): ')
            task_name = input()
            task_name = verifyName(task_name)


            print('Enter a event start time in this format HH:MM: ')
            time_start = input()
            time_start = verifyTime(time_start, duration, times)
            times.append(time_start)

            print('Enter a duration in this format HH:MM: ')
            duration = input()
            duration = verifyDuration(duration, time_start)


            tasks = Task(task_name,time_start,duration)

            pickle.dump(tasks, file)

            size+=1

            if duration_normal(duration, time_start):
                print('\nDo you want to continue input task?[y/n]: ')
                flag = input()
            else:
                flag = 'n'


    return size

def write_text(name_f: str, text):
    """
    Записуємо текст у файл
    """
    with open(name_f, "wb") as file:
        for i in range(len(text)):
            pickle.dump(text[i], file)

def next_event(text):
    min = 1440
    current_datetime = datetime.now()
    h = current_datetime.hour
    m = current_datetime.minute
    if h >= 10 and m < 10:
        print('Time now: ' + str(h) + ':' + '0' + str(m))
    elif h < 10 and m < 10:
        print('Time now: ' + '0' + str(h) + ':' + '0' + str(m))
    elif h < 10 and m >= 10:
        print('Time now: ' + '0' + str(h) + ':' + str(m))
    elif h >= 10 and m >= 10:
        print('Time now: ' + str(h) + ':' + str(m))
    dif = 0
    j = 0
    time_now = h * 60 + m
    for i in range(len(text)):
        ht = text[i].time_start.hours
        mt = text[i].time_start.minutes
        time_list = ht * 60 + mt
        dif = time_list - time_now
        if dif > 0 and dif < min:
            j = i
            min = dif
    if dif > 0:
        print('Next event: ' + text[j].name + ", " + text[j].time_start.toString() + ", " + "duration: " + text[j].duration.toString())
    else:
        print('Today you have no more business!!!')

def free_time(text, name_f):
    size = 0
    task_flag = False
    start = "13:00"
    finish = "23:59"
    free_dur = "10:59"
    file = open(name_f, "wb")
    for i in range(len(text)):
        time_n = text[i].time_start.hours * 60 + text[i].time_start.minutes
        dur_n = text[i].duration.hours * 60 + text[i].duration.minutes
        if time_n + dur_n >= 780:
            start = int_to_time(time_n + dur_n)
            if i + 1 < len(text):
                finish = int_to_time(text[i+1].time_start.hours * 60 + text[i+1].time_start.minutes)
            else:
                finish = '23:59'
            free_dur = int_to_time(time_to_int(finish) - time_to_int(start))
            free_text = FreeT(start, finish, free_dur)
            pickle.dump(free_text, file)
            task_flag = True
            size+=1
    if task_flag == False:
        free_text = FreeT(start, finish, free_dur)
        pickle.dump(free_text, file)

    return size

def read_file(name_f: str, size):
    """
    Читає текст з файлу
    """
    text = []
    print("Text on input:" + "\n")
    print(size)
    with open(name_f, "rb") as file:
        for i in range(size):
            text.append(pickle.load(file))
            str_t = text[i].name + ", " + text[i].time_start.toString() + ", " + "duration: " + text[i].duration.toString()
            print(str(str_t) + "\n")

    return text

def read_file2(name_f: str, size):
    """
    Читає текст з файлу
    """
    text = []
    print("Text on output:" + "\n")
    with open(name_f, "rb") as file:
        for i in range(size):
            text.append(pickle.load(file))
            str_t = text[i].start.toString() + "-" + text[i].finish.toString() + ", " + "duration: " + text[i].free_dur.toString()
            print(str(str_t) + "\n")



